### PR TITLE
Support specifying OSTree Parent and URL for creating upgrade commits

### PR DESCRIPTION
**Pull request by achilleas-k**
This PR changes the way we handle ostree parameters for compose requests.  The new behaviour is outlined in #2131.
The ostree parameter validation now happens in the ostree package.  The package defines custom error types for different error conditions.  This is necessary for the APIs to be able to communicate what went wrong.  Feedback on the error messages (and variable names) is very welcome.

I updated the weldr and cloud API tests to reflect the new changes and for higher coverage.  The tests can now set up a mock ostree repository to resolve the URL for testing successful compose requests.

The purpose of this change is to support creating upgrade commits (from one distro version to another).  Conventionally, the ostree `ref` depends on the os version (e.g., `rhel/8/x86_64/edge`).  To generate an upgrade commit (e.g., from RHEL 8.x to RHEL 9.0) we need to be able to specify the old ref (`rhel/8/...`) as the parent and the new ref for the new commit (`rhel/9/...`).  This is now supported, though we need to add integration tests for the specific use case.

-----

This pull request includes:

- [ ] adequate testing for the new functionality or fixed issue
- [ ] adequate documentation informing people about the change such as
  - [ ] submit a PR for the [guides](https://github.com/osbuild/guides) repository if this PR changed any behavior described there: https://www.osbuild.org/guides/

### CODE REVIEW

**Comment by henrywang** [comment]
@achilleas-k, my plan is to add different ref upgrade test, but not from rhel 8 to rhel 9 because that's not easy to do that in our CI. I'll cover rhel 8 to rhel 9 upgrade test on downstream CI. Any thoughts? Thanks.

**Comment by achilleas-k** [comment]
> @achilleas-k, my plan is to add different ref upgrade test, but not from rhel 8 to rhel 9 because that's not easy to do that in our CI. I'll cover rhel 8 to rhel 9 upgrade test on downstream CI. Any thoughts? Thanks.

Sounds good.  Testing the upgrade in downstream CI makes sense.
In our CI, we should at least test that we can build a commit based on a parent with a different ref.  I tested it locally last night when I opened the PR and I got a mysterious failure in the `org.osbuild.users` stage.  I'll look into it today.

**Comment by achilleas-k** [comment]
See https://github.com/osbuild/osbuild/pull/948 for user upgrade issue (and solution).
This is an existing issue that's not (directly) related to this PR, so I think we shouldn't block on this.
For testing this PR, we can temporarily work around it by not having a user (or not have a key for the user) in the upgrade commit.  This is what we already do in the `ostree-ng.sh` test.

**Comment by achilleas-k** [comment]
Needs a rebase and conflict resolution

**Comment by henrywang** [comment]
@achilleas-k Can we make a rebase because [PR#2219](https://github.com/osbuild/osbuild-composer/pull/2219) got merged already. So I can use new mock build to test this PR and add test case. Thanks!

**Comment by achilleas-k** [comment]
> @achilleas-k Can we make a rebase when you have time because [PR#2219](https://github.com/osbuild/osbuild-composer/pull/2219) got merged already. So I can use new mock build to test this PR and add test case. Thanks!

I'll rebase later today.  It's going to need some work because of the conflicting changes in the cloudapi from the Koji PR (#2214).

**Comment by achilleas-k** [comment]
Rebased

**Comment by henrywang** [comment]
@achilleas-k Looks openstack can't get configured RHEL 8.6 image.
```
Error: Error creating OpenStack server: Bad request with: [POST https://api.rhos-01.prod.psi.rdu2.redhat.com:13774/v2.1/servers], error message: {"badRequest": {"code": 400, "message": "Can not find requested image"}}
```

**Comment by henrywang** [comment]
@achilleas-k Tested on RHEL 9 (`RHEL 9 to RHEL 9 with different ref`)
```
[admin@vm-bios ~]$ rpm-ostree status
State: idle
Deployments:
● rhel-edge:test/rhel/x/x86_64/edge
                   Version: 9.0 (2022-02-03T08:31:56Z)
                    Commit: af35638feb86b1b6711de0bcb1315e3ca92ab7013eef3bde8ae7aa2c31868790

  rhel-edge:rhel/9/x86_64/edge
                   Version: 9.0 (2022-02-03T08:19:48Z)
                    Commit: 1b243056c63ffd14b0f7e1c1c78e91cef1ad553f30283938e2a8629c14789395
```

**Comment by henrywang** [comment]
@achilleas-k @gicmo RHEL 8.6 Edge system(`normal kernel 4.18.0-361.el8.x86_64`) upgrades to RHEL 9.0 Edge system (`rt kernel 5.14.0-51.rt21.51.el9.x86_64`)with `rebase` without any error. The issue is the upgrade blueprints can't have `[[customizations.user]]` configured. osbuild [PR #948](https://github.com/osbuild/osbuild/pull/948) will fix this issue.
```
[admin@localhost ~]$ rpm-ostree status
State: idle
Deployments:
* rhel-edge:rhel/9/x86_64/edge
                   Version: 9.0 (2022-02-03T13:39:07Z)
                    Commit: 6a470df710f6d65ee9332e15bf60e12c7561c99bb84fe9c0b60a43f811f44b49

  rhel-edge:rhel/8/x86_64/edge
                   Version: 8.6 (2022-02-01T06:08:47Z)
                    Commit: b28a859ba09fd3dcc45f63d0c8a49199db5b1d80747d2ebf43f1585686a210a4
```

**Comment by achilleas-k** [comment]
Update https://github.com/osbuild/osbuild-composer/pull/2201/commits/f9ca52ab47256868c8bcb3214c572cd126d8b6fa:
User stages (`org.osbuild.user`) for ostree commits are created without SSH keys anymore.  This avoids the problem with trying to write the SSH key in a non-existent home directory.  The key writing had no effect since the home directories created during a commit build are not retained.  Instead we create keys during first boot when the user's home directory is created.

@henrywang This should avoid the issue with the upgrade commit failing when there's a user customization with an SSH key.  Can you add this to the test?  I tested it locally for RHEL 8.6 before pushing.

**Comment by henrywang** [comment]
> Update [f9ca52a](https://github.com/osbuild/osbuild-composer/commit/f9ca52ab47256868c8bcb3214c572cd126d8b6fa): User stages (`org.osbuild.user`) for ostree commits are created without SSH keys anymore. This avoids the problem with trying to write the SSH key in a non-existent home directory. The key writing had no effect since the home directories created during a commit build are not retained. Instead we create keys during first boot when the user's home directory is created.
> 
> @henrywang This should avoid the issue with the upgrade commit failing when there's a user customization with an SSH key. Can you add this to the test? I tested it locally for RHEL 8.6 before pushing. Specifying a user that already exists in the parent while building an upgrade commit and having an SSH key now should append that SSH key to the existing `authorized_keys` file.

Thanks @achilleas-k for doing this. I tested in my local env, your fix works. And I added ostree rebase test case in this PR.

**Comment by achilleas-k** [comment]
OSTree rebase test is failing in CS8 and RHEL 9 at the failed boot / rollback test.  Wonder why 8.6 isn't.

**Comment by henrywang** [comment]
> OSTree rebase test is failing in CS8 and RHEL 9 at the failed boot / rollback test. Wonder why 8.6 isn't.

The CS8 issue should be related with cpu and memory resource on for CS8 vm. The cs8 vm(2 cpu, 4G memory) can't afford two vms running at the same time.
Resource after one VM got installed.
```
centos@centos-stream-8 ~]$ free -mh
              total        used        free      shared  buff/cache   available
Mem:          3.6Gi       1.8Gi       104Mi        10Mi       1.8Gi       1.6Gi
```
My `gitlab-ci-terraform` [PR #56](https://github.com/osbuild/gitlab-ci-terraform/pull/56) got merged. So we have large CS8 vm. Let's run test again on large cs8 vm.

**Comment by henrywang** [comment]
@achilleas-k YES! CS8 vm resource is definitely the root cause! It's green now with `large` vm. Really don't like this kind of debugging experience....

**Comment by henrywang** [comment]
The RHEL 9 issue is about greenboot. The UEFI vm can't rolllback and keeps rebooting after new deployment failed. This issue does not exist on BIOS vm. @runcom @nullr0ute Do you have any idea about this issue? Thanks.

**Comment by runcom** [comment]
> The UEFI vm can't rolllback and keeps rebooting after new deployment failed.

do we have any logs? Or manual testing with logs? Seems ostree related somehow or greenboot itself

**Comment by henrywang** [comment]
> > The UEFI vm can't rolllback and keeps rebooting after new deployment failed.
> 
> do we have any logs? Or manual testing with logs? Seems ostree related somehow or greenboot itself

How can I get log? The vm keeps rebooting, can't get into vm.
Here's the log I copied from console while vm booting.
```
[FAILED] Failed to start greenboot Health Checks Runner.
See 'systemctl status greenboot-healthcheck.service' for details.
[DEPEND] Dependency failed for Boot Completion Check.
[DEPEND] Dependency failed for Mark boot as successful in grubenv.
[DEPEND] Dependency failed for Multi-User System.
[DEPEND] Dependency failed for greenboot Success Scripts Runner.
         Starting greenboot Failure Scripts Runner...
         Starting Record Runlevel Change in UTMP...
[  OK  ] Finished greenboot Failure Scripts Runner.
[  OK  ] Started Modem Manager.
[  OK  ] Reached target Generic red boot target.
         Starting greenboot MotD Generator...
         Starting Reboot on red boot status...
```

**Comment by achilleas-k** [comment]
A (perhaps relevant?) note: this PR uses older repositories with greenboot < v14.  Any chance this is a bug fixed in the newer greenboot?

**Comment by runcom** [comment]
> A (perhaps relevant?) note: this PR uses older repositories with greenboot < 0.14. Any chance this is a bug fixed in the newer greenboot?

let's re-kick test on the other PR - rn they're just capped in the CI infra and aren't running :(

**Comment by henrywang** [comment]
@achilleas-k @runcom Then only failed/issue for this PR is [bz#2053469](https://bugzilla.redhat.com/show_bug.cgi?id=2053469)

**Comment by achilleas-k** [comment]
Rebased on main.
@henrywang I think after the rebase there might be some redundant checks in the ansible playbook.  Can you have a look please?  Thanks!

**Comment by runcom** [comment]
some RPM jobs failed in CI (might be infra related)

**Comment by henrywang** [comment]
@achilleas-k This PR needs your fix https://github.com/osbuild/osbuild-composer/pull/2142/commits/70f9780e9a67d57e8713466336284d5cc8977362 as well.

**Comment by henrywang** [comment]
@achilleas-k I just make a rebase because CS9 PR just got landed. I hope I did correct thing.

**Comment by henrywang** [comment]
@achilleas-k Still has rebooting loop issue. We can skip this test only on `ostree-rebase.sh` on UEFI test if there's no any fix. I made a change to support the skip.
```diff
diff --git i/test/cases/ostree-rebase.sh w/test/cases/ostree-rebase.sh
index 67a8af8a..2cb4a7af 100755
--- i/test/cases/ostree-rebase.sh
+++ w/test/cases/ostree-rebase.sh
@@ -601,7 +601,7 @@ ansible_ssh_common_args="-o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/
 EOF

 # Test IoT/Edge OS
-sudo ansible-playbook -v -i "${TEMPDIR}"/inventory -e image_type=rhel-edge -e ostree_commit="${UPGRADE_HASH}" -e ostree_ref="rhel-edge:${OSTREE_REF}" /usr/share/tests/osbuild-composer/ansible/check_ostree.yaml || RESULTS=0
+sudo ansible-playbook -v -i "${TEMPDIR}"/inventory -e image_type=rhel-edge -e ostree_commit="${UPGRADE_HASH}" -e ostree_ref="rhel-edge:${OSTREE_REF}" -e skip_rollback_test="true" /usr/share/tests/osbuild-composer/ansible/check_ostree.yaml || RESULTS=0
 check_result

 # Final success clean up
diff --git i/test/data/ansible/check_ostree.yaml w/test/data/ansible/check_ostree.yaml
index b0596714..5fb6ba9d 100644
--- i/test/data/ansible/check_ostree.yaml
+++ w/test/data/ansible/check_ostree.yaml
@@ -3,6 +3,7 @@
   become: no
   vars:
     workspace: "{{ lookup('env', 'WORKSPACE') }}"
+    skip_rollback_test: "false"
     total_counter: "0"
     failed_counter: "0"

@@ -535,6 +536,7 @@
         - name: failed count + 1
           set_fact:
             failed_counter: "{{ failed_counter | int + 1 }}"
+      when: skip_rollback_test == "false"

     # case: check ostree commit after rollback
     - name: check ostree commit after rollback
@@ -559,7 +561,9 @@
         - name: failed count + 1
           set_fact:
             failed_counter: "{{ failed_counter | int + 1 }}"
-      when: result_rollback is succeeded
+      when:
+        - skip_rollback_test == "false"
+        - result_rollback is succeeded

     # case: check greenboot* services log again
     - name: fallback log should be found here
@@ -588,7 +592,9 @@
         - name: failed count + 1
           set_fact:
             failed_counter: "{{ failed_counter | int + 1 }}"
-      when: result_rollback is succeeded
+      when:
+        - skip_rollback_test == "false"
+        - result_rollback is succeeded

     # case: check grubenv variables again
     - name: grubenv variables should contain boot_success=1
@@ -610,7 +616,9 @@
         - name: failed count + 1
           set_fact:
             failed_counter: "{{ failed_counter | int + 1 }}"
-      when: result_rollback is succeeded
+      when:
+        - skip_rollback_test == "false"
+        - result_rollback is succeeded

     - assert:
         that:
```

**Comment by achilleas-k** [comment]
> @achilleas-k Still has rebooting loop issue. We can skip this test only on `ostree-rebase.sh` on UEFI test if there's no any fix. I made a change to support the skip.

Thanks!  Applied.

**Comment by henrywang** [comment]
Thanks @achilleas-k! All ostree tests are GREEN! Re-tried two `integration` tests.

**Comment by teg** [comment]
> Thanks @achilleas-k! All ostree tests are GREEN! Re-tried two `integration` tests.

Looks like you are hitting #2322

**Comment by henrywang** [comment]
@teg OH! Ack. Thanks! Do we need wait for issue #2322 fixed before landing this PR? Thanks.

**Comment by teg** [comment]
I think not, but maybe @ondrejbudai or @achilleas-k have other thoughts

**Comment by achilleas-k** [comment]
I wouldn't be against an admin merge, but it's not my call :)

**Comment by teg** [comment]
I should amend: all for an admin pr, but let's make sure it doesn't conflict with the LVM one ...

**Comment by achilleas-k** [comment]
> I should amend: all for an admin pr, but let's make sure it doesn't conflict with the LVM one ...

Ah right.  I can check for that, or we can block until it's merged

**Comment by achilleas-k** [comment]
> I should amend: all for an admin pr, but let's make sure it doesn't conflict with the LVM one ...

Rebasing `gicmo/lvm` on this PR causes no conflicts and all local unit tests pass.  I can't think of any reason they might conflict beyond that.  The LVM PR is mostly contained in the `internal/disk` and `internal/osbuild*` packages and this one is API changes.

I understand if you want to wait to make absolutely sure though.

**Comment by teg** [comment]
Force pushed on top of LVM PR and various CI fixes.

**Comment by henrywang** [comment]
@achilleas-k @teg I just found two commits are missing after rebase. I added them back. :-)

**Comment by henrywang** [comment]
@achilleas-k Only `Integration: koji.sh, aws/centos-stream-9-x86_64` keeps fail after my twice re-tries.

**Comment by teg** [comment]
Oops! Thanks for the catch @henrywang

File: internal/distro/rhel86/pipelines.go
Comment (gicmo) [inline internal/distro/rhel86/pipelines.go:440]: In a follow up we could maybe have that as a utility function, maybe on the customisation object.

**Review by gicmo — APPROVED** [review]
This is great! Love it! Thanks so much @achilleas-k and @henrywang

**Pull request merged**
